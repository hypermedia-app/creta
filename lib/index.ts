import * as hb from 'hydra-box'
import type Api from 'hydra-box/Api'
import walk from '@fcostarodrigo/walk'
import { Debugger } from 'debug'

interface ApiInit {
  baseUri: string
  codePath: string
  defaultBase?: string
  apiPath: string
  log?: Debugger
}

export async function createApi({ apiPath, baseUri, codePath, defaultBase = 'urn:hydra-box:api', log }: ApiInit): Promise<Api> {
  const options: Api.Options = {
    path: '/api',
    codePath,
  }

  let api: Api | undefined
  for await (const file of walk(apiPath)) {
    if (!file.match(/\.ttl$/)) {
      continue
    }

    log && log(`Loading api from file ${file}`)
    if (api) {
      api = await api.fromFile(file)
    } else {
      api = await hb.Api.fromFile(file, options)
    }
  }

  if (!api) {
    throw new Error('No API files found')
  }

  api.rebase(defaultBase, baseUri)
  return api
}
