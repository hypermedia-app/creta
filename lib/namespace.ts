import namespace from '@rdfjs/namespace'

export const hex = namespace('http://hydra-ex.rest/vocab/')
export const hydraBox = namespace('http://hydra-box.org/schema/')
export const query = namespace('http://query.described.at/')
export const auth = namespace('http://auth.described.at/')
