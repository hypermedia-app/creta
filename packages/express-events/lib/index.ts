import { NamedNode } from 'rdf-js'
import clownface, { GraphPointer, MultiPointer } from 'clownface'
import { Activity } from '@rdfine/as'
import { DESCRIBE } from '@tpluscode/sparql-builder'
import { as, rdf, rdfs } from '@tpluscode/rdf-ns-builders'
import { sparql } from '@tpluscode/rdf-string'
import { code } from '@hydrofoil/namespaces'
import type * as express from 'express'
import $rdf from 'rdf-ext'
import { Handler, ns } from '../index'

export function isNamedNode(pointer: GraphPointer): pointer is GraphPointer<NamedNode> {
  return pointer.term.termType === 'NamedNode'
}

function hasOneImplementation(pointer: MultiPointer): pointer is GraphPointer {
  return !!pointer.term
}

function handlerQuery(event: Activity) {
  const activityTypes = [...event.types].map(({ id }) => id)

  return DESCRIBE`?handler`.WHERE`
  VALUES ?activity { ${as.Activity} ${activityTypes} }
      
  ${event.object ? sparql`${event.object.id} a ?type .` : sparql`VALUES ?type { ${rdfs.Resource} }`}

  ?handler a ${ns.EventHandler} ;
           ${ns.eventSpec} [
              ${rdf.predicate} ${rdf.type} ;
              ${rdf.object} ?activity ;
           ] ;
           ${ns.objectSpec} [
              ${rdf.predicate} ${rdf.type} ;
              ${rdf.object} ?type ;
           ] ;
           ${code.implementedBy} ?impl .`
}

export async function loadHandlers(req: express.Request, event: Activity) {
  const client = req.labyrinth.sparql
  const dataset = await $rdf.dataset().import(await handlerQuery(event).execute(client.query))

  return Promise.all(clownface({ dataset })
    .has(code.implementedBy)
    .toArray()
    .reduce((promises, handler) => {
      const implementedBy = handler.out(code.implementedBy)
      if (hasOneImplementation(implementedBy)) {
        const impl = req.loadCode<Handler>(implementedBy)
        if (!impl) {
          return promises
        }
        const promise = Promise.resolve().then(() => impl).then(impl => ({ handler, impl }))
        return [
          ...promises,
          promise,
        ]
      }

      req.knossos.log('No unique handler implementation found for handler %s', handler.value)
      return promises
    }, [] as Array<Promise<{ handler: GraphPointer; impl: Handler }>>))
}

export async function runHandler(handler: GraphPointer, impl: Handler, event: Activity, req: express.Request) {
  if (!impl) {
    req.knossos.log('Failed to load implementation of handler %s', handler.value)
    return
  }

  req.knossos.log('Running handler %s for event %s', impl.name, event.id.value)
  return impl({ event, req })
}
