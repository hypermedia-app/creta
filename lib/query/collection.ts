import { CONSTRUCT, SELECT } from '@tpluscode/sparql-builder'
import { hydra, ldp, rdf } from '@tpluscode/rdf-ns-builders'
import * as $rdf from '@rdfjs/data-model'
import { AnyPointer, GraphPointer } from 'clownface'
import { sparql, SparqlTemplateResult } from '@tpluscode/rdf-string'
import { IriTemplate, IriTemplateMapping } from '@rdfine/hydra'
import { Term, Variable } from 'rdf-js'
import { loaders } from '../rdfLoaders'
import { query } from '../namespace'

interface CreatePattern {
  (options: { subject: Variable; predicate: Term; object: AnyPointer }): string | SparqlTemplateResult
}

function createTemplateVariablePatterns(subject: Variable, queryPointer: AnyPointer, basePath: string) {
  return async (mapping: IriTemplateMapping): Promise<string | SparqlTemplateResult> => {
    const property = mapping.property

    if (hydra.pageIndex.equals(property.id)) {
      return ''
    }

    const value = queryPointer.out(property.id)
    if (value.values.length === 0) {
      return ''
    }

    const queryFilters = mapping.pointer.out(query.filter)
    if (!queryFilters.value) {
      return ''
    }

    const createPattern = await loaders.load<CreatePattern>(queryFilters.toArray()[0], { basePath })
    if (!createPattern) return ''

    return createPattern({
      subject,
      predicate: property.id,
      object: value,
    })
  }
}

function createManagesBlockPatterns(member: Variable) {
  return function (previous: SparqlTemplateResult, manages: GraphPointer): SparqlTemplateResult {
    const subject = manages.out(hydra.subject).term
    const predicate = manages.out(hydra.property).term
    const object = manages.out(hydra.object).term

    if (subject && predicate) {
      return sparql`${previous}\n${subject} ${predicate} ${member} .`
    }
    if (subject && object) {
      return sparql`${previous}\n${subject} ${member} ${object} .`
    }
    if (predicate && object) {
      return sparql`${previous}\n${member} ${predicate} ${object} .`
    }

    return previous
  }
}

type SelectBuilder = ReturnType<typeof SELECT>

function createOrdering(api: GraphPointer, collection: GraphPointer, subject: Variable): { patterns: SparqlTemplateResult; addClauses(q: SelectBuilder): SelectBuilder } {
  const orders = api.node(collection.out(rdf.type) as any).out(query.order).toArray()
  if (!orders.length) {
    return {
      patterns: sparql``,
      addClauses: q => q,
    }
  }

  let orderIndex = 0
  let patterns = sparql``
  const clauses: Array<{ variable: Variable; descending: boolean }> = []

  for (const order of orders[0].list()!) {
    const propertyPath = order.out(query.path).list()
    if (!propertyPath) continue

    const path = [...propertyPath].reduce((current, prop, index) => {
      const next = index ? sparql`/${prop.term}` : sparql`${prop.term}`

      return sparql`${current}${next}`
    }, sparql``)
    const variable = $rdf.variable(`order${++orderIndex}`)

    const pattern = sparql`OPTIONAL { ${subject} ${path} ${variable} } .`
    patterns = sparql`${patterns}\n${pattern}`

    clauses.push({
      variable,
      descending: ldp.Descending.equals(order.out(query.direction).term),
    })
  }

  return {
    patterns,
    addClauses(query) {
      return clauses.reduce((orderedQuery, { variable, descending }) => {
        return orderedQuery.ORDER().BY(variable, descending)
      }, query)
    },
  }
}

export async function getMemberQuery(api: GraphPointer, collection: GraphPointer, query: AnyPointer, variables: IriTemplate | null, pageSize: number, basePath: string) {
  const subject = $rdf.variable('member')
  const managesBlockPatterns = collection.out(hydra.manages).toArray().reduce(createManagesBlockPatterns(subject), sparql``)
  let filterPatters: Array<string | SparqlTemplateResult> = []
  if (variables) {
    filterPatters = await Promise.all(variables.mapping.map(createTemplateVariablePatterns(subject, query, basePath)))
  }

  const order = createOrdering(api, collection, subject)

  const memberPatterns = sparql`${managesBlockPatterns}\n${filterPatters}`

  let subselect = SELECT`?g`.WHERE` 
              GRAPH ?g {
                ${memberPatterns}
                
                ${order.patterns}
              }`

  if (variables && variables.mapping.some(mapping => mapping.property.equals(hydra.pageIndex))) {
    const page = Number.parseInt(query.out(hydra.pageIndex).value || '1')
    subselect = subselect.LIMIT(pageSize).OFFSET((page - 1) * pageSize)
    subselect = order.addClauses(subselect)
  }

  return {
    members: CONSTRUCT`?s ?p ?o. ?is ?io ?ip`.WHERE`
        {
            ${subselect}
        }
        
        GRAPH ?g { ?s ?p ?o }`,
    totals: SELECT`(count(${subject}) as ?count)`.WHERE`GRAPH ?g { ${memberPatterns} }`,
  }
}
