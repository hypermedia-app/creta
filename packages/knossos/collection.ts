import { NamedNode } from 'rdf-js'
import { Router } from 'express'
import asyncMiddleware from 'middleware-async'
import { hydra, rdf } from '@tpluscode/rdf-ns-builders'
import { created } from '@hydrofoil/knossos-events/activity'
import error from 'http-errors'
import httpStatus from 'http-status'
import $rdf from 'rdf-ext'
import { fromPointer } from '@rdfine/hydra/lib/IriTemplate'
import { IriTemplateBundle } from '@rdfine/hydra/bundles'
import RdfResource, { ResourceIdentifier } from '@tpluscode/rdfine'
import clownface, { AnyPointer, GraphPointer } from 'clownface'
import { shaclValidate } from './shacl'
import { knossos } from './lib/namespace'
import { save } from './lib/resource'

RdfResource.factory.addMixin(...IriTemplateBundle)

function checkMemberTemplate(ptr: AnyPointer): ptr is GraphPointer<ResourceIdentifier> {
  return ptr.term?.termType === 'NamedNode' || ptr.term?.termType === 'BlankNode'
}

function rename(member: GraphPointer, id: NamedNode): GraphPointer<NamedNode> {
  for (const match of member.dataset.match(member.term)) {
    member.dataset.delete(match)
    member.dataset.add($rdf.quad(id, match.predicate, match.object, match.graph))
  }
  for (const match of member.dataset.match(null, null, member.term)) {
    member.dataset.delete(match)
    member.dataset.add($rdf.quad(match.subject, match.predicate, id, match.graph))
  }

  return member.node(id)
}

export const POSTCreate = Router().use(shaclValidate).use(asyncMiddleware(async (req, res, next) => {
  const api = clownface(req.hydra.api)
  const collection = await req.hydra.resource.clownface()
  const types = collection.out([hydra.manages, hydra.memberAssertion]).has(hydra.property, rdf.type).out(hydra.object)

  if (!types.terms.length) {
    return next(new error.InternalServerError('Collection does not have a member assertion with `hydra:property rdf:type`'))
  }

  const { type } = rdf
  const { memberTemplate } = knossos
  const memberTemplateS = api.node(collection.out(type)).out(memberTemplate)
  if (!checkMemberTemplate(memberTemplateS)) {
    req.knossos.log.extend('collection')('Found member templates %o', memberTemplateS.map(mt => mt.out(hydra.template).value))
    return next(new error.InternalServerError(`No unique knossos:memberTemplate found for collection ${collection.value}`))
  }

  let member = await req.resource()
  const memberId = $rdf.namedNode(new URL(fromPointer(memberTemplateS).expand(member), req.absoluteUrl()).toString())

  if (await req.knossos.store.exists(memberId)) {
    return next(new error.Conflict())
  }

  req.knossos.log(`Creating resource ${memberId.value}`)
  member = rename(member, memberId).addOut(rdf.type, types)

  await save({ resource: member, req })

  res.event(created(member, {
    context: req.hydra.resource.term,
  }))

  await res.event.handleImmediate()

  res.status(httpStatus.CREATED)
  res.setHeader('Location', memberId.value)
  return res.resource(await req.knossos.store.load(member.term))
}))
