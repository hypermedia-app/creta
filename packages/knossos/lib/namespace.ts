import namespace from '@rdfjs/namespace'

type KnossosTerms = 'createWithPUT' | 'memberTemplate' | 'SystemAccount'

export const knossos = namespace<KnossosTerms>('https://hypermedia.app/knossos#')
