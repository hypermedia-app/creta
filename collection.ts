import asyncMiddleware from 'middleware-async'
import clownface, { AnyPointer, GraphPointer, MultiPointer } from 'clownface'
import $rdf from 'rdf-ext'
import { ResourceIdentifier } from '@tpluscode/rdfine'
import { Request } from 'express'
import { hydra, rdf } from '@tpluscode/rdf-ns-builders'
import { IriTemplate, IriTemplateMixin } from '@rdfine/hydra'
import { hydraBox, query } from './lib/namespace'
import { getSparqlQuery } from './lib/query/collection'
import { loadLinkedResources } from './lib/query/eagerLinks'
import { protectedResource } from './resource'

const pageSize = 12
const emptyDataset = clownface({ dataset: $rdf.dataset() })

function templateParamsForPage(query: AnyPointer, page: number) {
  const clone = clownface({ dataset: $rdf.dataset([...query.dataset]) })
    .in().toArray()[0]

  if (!clone) {
    return clownface({ dataset: $rdf.dataset() }).blankNode().addOut(hydra.pageIndex, page)
  }

  if (page === 1) {
    return clone.deleteOut(hydra.pageIndex)
  }

  return clone.deleteOut(hydra.pageIndex).addOut(hydra.pageIndex, page)
}

function getTemplate(req: Request): IriTemplate | null {
  const templateVariables = req.hydra.operation.out(hydraBox.variables) as MultiPointer<ResourceIdentifier>
  if (templateVariables.term) {
    return new IriTemplateMixin.Class(templateVariables.toArray()[0])
  }

  return null
}

function addTemplateMappings(collection: GraphPointer, template: IriTemplate, request: AnyPointer = emptyDataset): void {
  collection.addOut(query.templateMappings, currMappings => {
    template.mapping.forEach(mapping => {
      const property = mapping.property.id
      currMappings.addOut(property, request.out(property))
    })
  })
}

function addCollectionViews(collection: GraphPointer, total: number, template: IriTemplate, request: AnyPointer = emptyDataset): void {
  if (!template.mapping.some(m => m.property.equals(hydra.pageIndex))) {
    return
  }

  collection
    .addOut(hydra.view, view => {
      const pageIndex = Number.parseInt(request.out(hydra.pageIndex).value || '1')
      const totalPages = Math.floor(total / pageSize) + 1

      view.addOut(rdf.type, hydra.PartialCollectionView)
      view.addOut(hydra.first, $rdf.namedNode(template.expand(templateParamsForPage(request, 1))))
      view.addOut(hydra.last, $rdf.namedNode(template.expand(templateParamsForPage(request, totalPages))))
      if (pageIndex > 1) {
        view.addOut(hydra.previous, $rdf.namedNode(template.expand(templateParamsForPage(request, pageIndex - 1))))
      }
      if (pageIndex < totalPages) {
        view.addOut(hydra.next, $rdf.namedNode(template.expand(templateParamsForPage(request, pageIndex + 1))))
      }
    })
}

export const get = protectedResource(asyncMiddleware(async (req, res) => {
  const dataset = $rdf.dataset([...req.hydra.resource.dataset])
  const collection = clownface({ dataset }).namedNode(req.hydra.resource.term)

  let request: AnyPointer | undefined
  if (req.dataset) {
    request = clownface({ dataset: await req.dataset() })
  }

  const includeLinked = req.hydra.operation.out(query.include)

  const template = getTemplate(req)

  const pageQuery = await getSparqlQuery({
    api: clownface(req.hydra.api),
    collection,
    query: request,
    variables: template,
    pageSize,
    basePath: req.hydra.api.codePath,
  })

  let total = 0
  if (pageQuery) {
    const page = await pageQuery.members.execute(req.app.sparql.query)
    const totals = pageQuery.totals.execute(req.app.sparql.query)
    for await (const result of await totals) {
      total = Number.parseInt(result.count.value)
    }
    await collection.dataset.import(page)
  }
  collection.addOut(hydra.totalItems, total)

  collection.namedNode(req.hydra.resource.term)
    .addOut(hydra.member, clownface({ dataset }).has(rdf.type, collection.out(hydra.manages).has(hydra.property, rdf.type).out(hydra.object)))

  dataset.addAll(await loadLinkedResources(collection.out(hydra.member), includeLinked, req.app.sparql))

  if (template) {
    addTemplateMappings(collection, template, request)
    addCollectionViews(collection, total, template, request)
  }

  res.setLink(req.hydra.resource.term.value, 'canonical')
  return res.dataset(dataset)
}))
