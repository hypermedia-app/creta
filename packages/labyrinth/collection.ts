import { RequestHandler, Router, Response } from 'express'
import clownface from 'clownface'
import { knossos } from '@hydrofoil/vocabularies/builders/strict'
import asyncMiddleware from 'middleware-async'
import { hydra } from '@tpluscode/rdf-ns-builders'
import { preprocessMiddleware } from './lib/middleware/preprocessResource'
import * as lib from './lib/collection'

export type CollectionResponse = Response<any, Partial<lib.CollectionLocals>>

function assertLocals<T extends keyof lib.CollectionLocals>(locals: Partial<lib.CollectionLocals>, ...keys: T[]): locals is Pick<lib.CollectionLocals, T> {
  return keys.every(key => key in locals)
}

const sendResponse: RequestHandler = (req, res) => {
  const { dataset } = res.locals.collection
  res.setLink(req.hydra.resource.term.value, 'canonical')
  return res.dataset(dataset)
}

export const get = Router()
  .use(asyncMiddleware(async (req, res: CollectionResponse, next) => {
    res.locals = await lib.loadCollection(req)
    next()
  }))
  .use(asyncMiddleware(async (req, res: CollectionResponse, next) => {
    if (!assertLocals(res.locals, 'collection', 'queryParams', 'pageSize', 'search', 'searchTemplate')) {
      return next(new Error('Collection not loaded'))
    }

    res.locals = {
      ...res.locals,
      ...await lib.initQueries(res.locals, req),
    }
    next()
  }))
  .use(asyncMiddleware(async (req, res: CollectionResponse, next) => {
    if (!assertLocals(res.locals, 'collection', 'queries')) {
      return next(new Error('Collection not loaded'))
    }

    const {
      members, total, memberData,
    } = await lib.runQueries(res.locals)

    await res.locals.collection.dataset.import(memberData)
    res.locals.collection.addOut(hydra.member, [...members.values()])
    res.locals.collection.deleteOut(hydra.totalItems).addOut(hydra.totalItems, total)

    next()
  }))
  .use((req, res: CollectionResponse, next) => {
    if (!assertLocals(res.locals, 'collection', 'searchTemplate', 'queryParams', 'pageSize', 'total')) {
      return next(new Error('Collection not loaded'))
    }

    const views = lib.createViews(res.locals)

    if (views) {
      res.locals.collection.addOut(hydra.view, [...views.terms])
      res.locals.collection.dataset.addAll([...views.dataset])
    }
    next()
  })
  .use(preprocessMiddleware({
    async getResource(req, res: CollectionResponse) {
      const { dataset } = res.locals.collection!
      return clownface({ dataset, term: req.hydra.resource.term })
    },
    predicate: knossos.preprocessResponse,
  }))
  .use(sendResponse)
