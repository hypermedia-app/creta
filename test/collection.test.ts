import { describe, it } from 'mocha'
import { expect } from 'chai'
import express from 'express'
import request from 'supertest'
import { hydraBox } from './support/hydra-box'
import { get } from '../collection'
import { auth } from '../lib/namespace'
import { ex } from './support/namespace'

describe('labyrinth/collection', () => {
  describe('get', () => {
    it('returns 403 when collection is restricted', async () => {
      // given
      const app = express()
      app.use(hydraBox({
        setup: hydra => {
          hydra.operation.addOut(auth.required, true)
        },
      }))
      app.use(get)

      // when
      const response = request(app).get('/')

      // then
      await response.expect(403)
    })

    it('sets canonical link header', async () => {
      // given
      const app = express()
      app.use(hydraBox({
        setup: hydra => {
          hydra.resource.term = ex.people
        },
      }))
      app.use(get)

      // when
      const { links } = await request(app).get('/')

      // then
      expect(links).to.have.property('canonical', 'http://example.com/people')
    })
  })
})
