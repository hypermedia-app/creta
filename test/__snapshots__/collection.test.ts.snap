exports["labyrinth/collection get returns empty collection when no query is returned"] = [
	{
		"@id": "http://example.com/movies",
		"@type": "http://example.com/Collection"
	},
	{
		"@id": "http://example.com/movies",
		"http://www.w3.org/ns/hydra/core#totalItems": {
			"@type": "http://www.w3.org/2001/XMLSchema#integer",
			"@value": "0"
		}
	}
];

